### PR TITLE
2884 preparser errors multi record rows wrong length missing space filled m3 tribal t3

### LINKED ISSUE #2884: Generate preparser errors when multi-record rows are the wrong length or are missing space-filled second records - M3 and Tribal T3
**Description:**

As a follow up to ticket [#2757](https://github.com/raft-tech/TANF-app/issues/2757) (similar issue for T3), this ticket addresses cases for M3 and Tribal T3.

**Acceptance Criteria:** 
_Create a list of functional outcomes that must be achieved to complete this issue_

- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] Create new validator for M3 and Tribal T3 to be used in pre-validation for first and second child as per tables in [#2757](https://github.com/raft-tech/TANF-app/issues/2757) (in which T3 was done)
- [ ] Create associated tests
- [ ] _Task 3_
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- _Note 1_
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by raftmsohani**
## Summary of Changes

Pull request closes #2884 _

Continued changes to Tribal T3 and M3 to count for space filled second child cases. If the second child is truncated then parser generates pre-parse error.

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [x] tribal t3 and ssp m3 records according to spec
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py [hunk 1]
```diff
@@ -7,13 +7,14 @@
...
 
+FIRST_CHILD = 1
+SECOND_CHILD = 2
 
 child_one = RowSchema(
     record_type="T3",
     document=Tribal_TANF_T3DataSubmissionDocument(),
     preparsing_validators=[
-        validators.notEmpty(start=19, end=60),
-        validators.caseNumberNotEmpty(8, 19),
```
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:16]: I think this was accidentally removed?

File: tdrs-backend/tdpservice/parsers/validators.py [hunk 2]
```diff
@@ -710,6 +721,36 @@ def validate_header_rpt_month_year(datafile, header, generate_error):
...
+        if not _is_empty(value, 1, 60) and len(value) >= 60:
+            return (True, None)
+        else:
+            return (False, "1st child record truncated.")
+
+    def t3_second_child_validator_func(value, temp, friendly_name, item_num):
+        if not _is_empty(value, 60, 101) and len(value) >= 101 and not _is_empty(value, 8, 19):
+            return (True, None)
+        else:
+            return (False, "2nd child record truncated.")
```
Comment (elipe17) [inline tdrs-backend/tdpservice/parsers/validators.py]: I can't remember if I asked in the OG PR, but is this the error language Alex and Miles provided?

**Review by elipe17 — COMMENTED** [review]
Couple of questions

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py (hunk 1 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:16]: Technically this is being validated in the new child validator now

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 2 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/validators.py]: no, I pinged Alex and Miles in the other PR for review

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py (hunk 1 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:16]: I don't think thats true. That validator checks if the the entire range is empty. If that range has one valid character in it the validator will pass right? Also, even if it did fail the user doesn't know that their case number is malformed.

Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:16]: you are correct, but my discussion with Alex on this was that she wanted these to be parsed if and only if at least one of the fields are filled (doesn't mater which). So even if the case_number is empty but another field is filled, then parse. Having both these validators will stop this.
But I also see the point: if the case_number is empty, maybe we have to validate that separately. @ADPennington : what do you think?

File: tdrs-backend/tdpservice/parsers/test/test_parse.py [hunk 3]
```diff
@@ -1555,6 +1550,50 @@ def test_parse_tribal_section_4_file(tribal_section_4_file, dfs):
...
+    dfs.datafile = file_fixture
+    parse.parse_datafile(file_fixture, dfs)
+    errors = ParserError.objects.all()
+    for i in errors:
+        logger.info('________________________' + i.error_message)
+    t3 = TANF_T3.objects.all()
+    assert t3.exists() == result
+
+    parser_errors = ParserError.objects.all()
+    assert parser_errors.count() == number_of_errors
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/test/test_parse.py:1721]: can the error messages be asserted on as well?

File: tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py (hunk 1 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/schema_defs/tribal_tanf/t3.py:16]: The notEmpty condition is handled in ```validators.t3_m3_child_validator```

File: tdrs-backend/tdpservice/parsers/test/test_parse.py [hunk 4]
```diff
@@ -1555,6 +1550,137 @@ def test_parse_tribal_section_4_file(tribal_section_4_file, dfs):
...
+                          ('two_child_second_filled', True, 9, 'T3: Year 6    must be larger than 1900.')])
+@pytest.mark.django_db()
+def test_misformatted_multi_records(file_fixture, result, number_of_errors, error_message, request, dfs):
+    """Test that (not space filled) multi-records are caught."""
+    file_fixture = request.getfixturevalue(file_fixture)
+    dfs.datafile = file_fixture
+    parse.parse_datafile(file_fixture, dfs)
+    parser_errors = ParserError.objects.filter(file=file_fixture)
+    t3 = TANF_T3.objects.all()
+    assert t3.exists() == result
```
Comment (jtimpe) [inline tdrs-backend/tdpservice/parsers/test/test_parse.py]: another suggestion is to assert on the number of records created (1 or 2 depending on the case)

**Review by jtimpe — APPROVED** [review]
minor test suggestion. approved pending input from alex

File: tdrs-backend/tdpservice/parsers/test/test_parse.py (hunk 4 above)
Comment (elipe17) [reply tdrs-backend/tdpservice/parsers/test/test_parse.py]: I like the idea of asserting on the number of records versus their existence. I feel like that could save some potential future headaches if/when we start refactoring the parser.

**Review by elipe17 — COMMENTED** [review]
One other question. Should the base branch for this PR be `2757-generate-preparser-errors-multi-record-rows-wrong-length` or does it not matter since they have the ability to be merged independently?

File: tdrs-backend/tdpservice/parsers/validators.py (hunk 2 above)
Comment (raftmsohani) [reply tdrs-backend/tdpservice/parsers/validators.py]: Fixed the language,

**Review by elipe17 — APPROVED** [review]
Approving so @ADPennington can comment on the removal of that validator we discussed.

**Comment by ADPennington** [comment]
@raftmsohani please see unit test findings here: https://app.circleci.com/pipelines/github/raft-tech/TANF-app/23996/workflows/0c4bb4a5-8229-4329-8721-536980f32f25/jobs/71070

**Comment by raftmsohani** [comment]
> @raftmsohani please see unit test findings here: https://app.circleci.com/pipelines/github/raft-tech/TANF-app/23996/workflows/0c4bb4a5-8229-4329-8721-536980f32f25/jobs/71070

@ADPennington I am working on it 👍

**Comment by raftmsohani** [comment]
> > @raftmsohani please see unit test findings here: https://app.circleci.com/pipelines/github/raft-tech/TANF-app/23996/workflows/0c4bb4a5-8229-4329-8721-536980f32f25/jobs/71070
> 
> @ADPennington I am working on it 👍

Resolved

**Review by ADPennington — APPROVED** [review]
lgtm @raftmsohani 🚀

**Pull request merged**